File: src/java/main/commands/RichJson_merge.java
package commands;

import core.*;
import helper.RichJsonHelper;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class RichJson_merge implements RichJsonCommand {
    @Override
    @SuppressWarnings("unchecked")
    public Object execute(RichJsonParser parser, RichJsonContext context) throws Exception {
        String[] refs = ((String) context.currentMember).split(RichJsonConstants.COMMAND_DELIMITER);
        Object structOrArray = parser.cache.stack.get(context.currentAddress);
        String originalAddress = context.currentAddress;

        context.currentMember = refs[0].trim();
        context.currentMember = RichJsonCommandHolder.executeCommand("ref", parser, context);
        context.currentAddress = originalAddress;

        if (context.currentMember instanceof Map) {
            Map<String, Object> targetMap = (Map<String, Object>) structOrArray;
            RichJsonHelper.mergeIntoTarget(targetMap, (Map<String, Object>) context.currentMember);

            for (int i = 1; i < refs.length; i++) {
                context.currentMember = refs[i].trim();
                Object nextRef = RichJsonCommandHolder.executeCommand("ref", parser, context);
                context.currentAddress = originalAddress;
                RichJsonHelper.mergeIntoTarget(targetMap, (Map<String, Object>) nextRef);
            }
            return targetMap;
        } else {
            List<Object> targetList = new ArrayList<>();
            parser.cache.stack.put(originalAddress, targetList);

            for (int i = 0; i < refs.length; i++) {
                context.currentMember = refs[i].trim();
                List<Object> nextRef = (List<Object>) RichJsonCommandHolder.executeCommand("ref", parser, context);
                context.currentAddress = originalAddress;
                targetList.addAll(nextRef);
            }
            return targetList;
        }
    }
}

File: src/c#/main/commands/RichJson_merge.cs
using System;
using System.Collections.Generic;
using main.core;
using main.helper;

namespace main.commands
{
    public class RichJson_merge
    {
        public object Execute(RichJsonParser parser, RichJsonContext context)
        {
            string[] refs = ((string)context.CurrentMember).Split(new[] { RichJsonConstants.COMMAND_DELIMITER }, StringSplitOptions.None);
            object structOrArray = parser.Cache.Stack[context.CurrentAddress];
            string originalAddress = context.CurrentAddress;

            context.CurrentMember = refs[0].Trim();
            context.CurrentMember = RichJsonCommandHolder.ExecuteCommand("ref", parser, context);
            context.CurrentAddress = originalAddress;

            if (context.CurrentMember is IDictionary<string, object>)
            {
                var targetMap = (IDictionary<string, object>)structOrArray;
                RichJsonHelper.MergeIntoTarget(targetMap, (IDictionary<string, object>)context.CurrentMember);

                for (int i = 1; i < refs.Length; i++)
                {
                    context.CurrentMember = refs[i].Trim();
                    object nextRef = RichJsonCommandHolder.ExecuteCommand("ref", parser, context);
                    context.CurrentAddress = originalAddress;
                    RichJsonHelper.MergeIntoTarget(targetMap, (IDictionary<string, object>)nextRef);
                }
                return targetMap;
            }
            else
            {
                var targetList = new List<object>();
                parser.Cache.Stack[originalAddress] = targetList;

                for (int i = 0; i < refs.Length; i++)
                {
                    context.CurrentMember = refs[i].Trim();
                    var nextRef = (System.Collections.IList)RichJsonCommandHolder.ExecuteCommand("ref", parser, context);
                    context.CurrentAddress = originalAddress;

                    foreach (var item in nextRef)
                    {
                        targetList.Add(item);
                    }
                }
                return targetList;
            }
        }
    }
}